README.md
# MyExpense Web Application

## Détails du challenge
* **Difficulté :** Easy
* **Type :** Réaliste
* **Technologies :** PHP / MySQL

MyExpense est une application **volontairement vulnérable** permettant de s’entraîner à détecter et à exploiter différentes vulnérabilités web. Contrairement à une application de type "challenge" plus classique (qui permet de s'entrainer sur une seule vulnérabilité précise), MyExpense contient un ensemble de vulnérabilités. 

Le second objectif de cette application est d'immerger l'attaquant dans un environnement professionnel. Pour cela, un effort a été fourni afin que l'application ressemble le plus possible à un outil destiné à une utilisation interne par les employés de l'entreprise "Futura Business Informatique". De plus, il est possible (et nécessaire afin de récupérer le flag/drapeau validant le challenge) de lancer des scripts permettant de simuler l'utilisation de l'application par des employés afin d'exploiter certaines vulnérabilités ayant pour victime les utilisateurs authentifiés.

## Point d'attention
L'application étant volontairement vulnérable il n'est pas souhaitable de l'exposer sur le réseau Internet car d'autres personnes que vous pourront y y accéder. Il est conseillé d'utiliser une machine virtuelle (en utilisant par exemple le logiciel [VirtualBox](https://www.virtualbox.org/)) et de restreindre la connectivité **hôte/machine vulnérable** (mode Réseau privé hôte).

A des fins d’entraînement, il est conseillé de ne pas utiliser d'outils de détection ou d'exploitation de vulnérabilités (scanner de vulnérabilité, etc) et de ne pas regarder le code source de l'application (mode *blackbox*).  

## Scénario
Afin de simuler une attaque ayant un but et une motivation réelle, voici un scénario menant à la validation du challenge :
```
Vous vous nommez "Samuel Lamotte" et vous venez de vous faire licencier par votre entreprise "Furtura Business Informatique". 
Malheureusement à cause de votre départ précipité, vous n'avez pas eu le temps de valider votre note de frais de votre dernier déplacement professionnel, note s'élevant quand même à 750 € correspondant à un aller retour en avion pour une mission chez votre dernier client. 
Par peur que votre ex-employeur ne veuille vous rembourser cette note de frais sur simple demande, vous avez décidé de pirater l'application interne nommé **"MyExpense"** destinée à gérer les remboursements des notes de frais des employés.

Vous êtes donc dans votre voiture, sur le parking de l'entreprise voisine et êtes connecté au Wi-Fi interne de l'entreprise (la clé n'a toujours pas été changée après votre départ).
L'application est protégée par une authentification de type nom d'utilisateur/mot de passe et vous espérez que l'administrateur n'a pas encore modifié ou supprimé votre accès.

Vos identifiants étaient : slamotte/fzghn4lw

Une fois le challenge réussi, le flag sera affiché sur l'application en étant connecté avec votre compte.

A vos claviers !
```
## Téléchargement de la machine virtuelle
Il est plus simple de récupérer l'application en téléchargeant directement la machine virtuelle au format _.vbox_ :
[MyExpense Virtual Machine](http://www.mediafire.com/file/rzg9dxt14tm4ozs/MyExpense_%28192.168.56.140%29.ova)

Par défaut, l'application écoute sur l'adresse IP 192.168.56.140 et est accessible via le nom de domaine _myexpense.fbi.com_. Afin que cela fonctionne sur les machines d'attaques, il faut éditer les fichiers hosts des différentes machines utilisées et y ajouter la ligne suivante :
```
192.168.56.140  myexpense.fbi.com
```
**Note :** Sous linux le fichier se situe dans _/etc/hosts_ et sous Windows dans _C:\Windows\System32\drivers\etc\hosts_

## Réinitialiser la base de données de l'application
Il est possible de restaurer la base de données de l'application afin de pouvoir recommancer à partir de l'état initial. Pour cela il faut se rendre sur l'url _http://myexpense.fbi.com/config/setup.php_ puis de cliquer sur _Create/restore the database_. Un message indiquant que l'opération a été effectuée avec succès doit apparaitre :
![Création BDD](https://github.com/Sharpforce/Web/blob/master/.images/04.%20Challenges/01.%20MyExpense_config_db_success.png)

## Installation à partir des sources

### Système d'exploitation
L'installation a été testée sous un système d'exploitation Linux Debian9.

### Installation des paquets
Dans un premier temps il est nécessaire d'installer les paquets correspondant au serveur web Apache, à PHP ainsi qu'a la base de données MySql (qui sera MariaDB pour une Debian9) :
```
# apt-get install apache2 mysql-server php7.0 php7.0-mysql
```

Puis :
```
# rm /var/www/html/index.html
```

### Installation de git et récupération du code source
Il est possible d'installer l'utilitaire git afin de récupérer les sources de l'application ou alors de télécharger directement sur GitHub l'archive au format ZIP : [MyExpense.zip](https://github.com/Sharpforce/Web/archive/MyExpense-master.zip).

#### Par l'utilitaire Git
```
# apt-get install git
# cd /tmp
# git clone -b MyExpense-master https://github.com/Sharpforce/Web-Security.git
# rm -rf /tmp/Web-Security/.git
```

Il faut ensuite déplacer le code source au sein du répertoire */var/www/html/* d'Apache :
```
# mv /tmp/Web-Security/* /var/www/html/
```

#### Par l'archive Zip
Il sera peut être nécessaire d'installer le paquet **unzip** afin d'extraire l'archive :
```
# apt-get install unzip
```

Puis extraire l'archive :
```
# cd /tmp
# wget https://github.com/Sharpforce/Web-Security/archive/MyExpense-master.zip
# unzip MyExpense-master.zip
# mv Web-Security-MyExpense-master/* /var/www/html
# mv Web-Security-MyExpense-master/.htaccess /var/www/html/.htaccess
```

### Configuration Apache2
Une modification doit être effectuée au sein du fichier de configuration d'Apache2 afin de prendre en compte le fichier _.htacess_ :
```
# vim /etc/apache2/apache2.conf
```

Modifier la ligne _AllowOverride None_ par _AllowOverride All_ dans la partie _<Directory /var/www/>_ :
```
<Directory /var/www/>
        Options Indexes FollowSymLinks
        AllowOverride All
        Require all granted
</Directory>
```

Puis redémarrer le service Apache2 :
```
# service apache2 restart
```

### Configuration de la base de données
```
# mysql -u root
```

Création d'un nouvel utilisateur et attribution des droits :
```
MariaDB [(none)]> grant all on *.* to MyExpenseUser@localhost identified by 'password';
Query OK, 0 rows affected (0.00 sec)

MariaDB [(none)]> flush privileges;
Query OK, 0 rows affected (0.00 sec)

MariaDB [(none)]> quit
Bye
```

Il faut maintenant renseigner ces informations au niveau du fichier de configuration de l'application :
```
# vim /var/www/html/config/config.inc.php
```

Puis modifier les informations de connexion :
```
  // Database Configuration
  $_bdd = array();
  $_bdd['server'] = "127.0.0.1";
  $_bdd['port'] = "3306";
  $_bdd['user'] = "MyExpenseUser";
  $_bdd['password'] = "password";
  $_bdd['database'] = "myexpense";
```

### Création de la base de données
La configuration de l'application MyExpense doit être maintenant accessible via l'url _http://192.168.56.141/config/setup.php_ :
![Configuration BDD](https://github.com/Sharpforce/Web/blob/master/.images/04.%20Challenges/01.%20MyExpense_config_db.png)

Vérifier les informations puis cliquer sur **Create/Restore the database** :
![Création BDD](https://github.com/Sharpforce/Web/blob/master/.images/04.%20Challenges/01.%20MyExpense_config_db_success.png)

### Installation des scripts employés
L'application doit être maintenant installée et fonctionnelle. Afin de pouvoir compléter le challenge proposé et de rendre l'expérience un peu plus immersif, il est nécessaire d'installer les scripts de simulation d'action des employés.

Tout d'abord, renseigner le fichier hosts (_/etc/hosts_) sur la machine virtuelle **ET** sur votre machine d'attaque comme indiqué ci-dessous :
```
# vim /etc/hosts
```

Ici _192.168.56.100_ est un exemple, il faut insérer l'adresse IP de la machine virtulle qui héberge l'application :
```
192.168.56.100  myexpense.fbi.com
```
**Note :** Ne pas oublier de modifier le fichier hosts également sous votre machine Windows par exemple (_C:\Windows\System32\drivers\etc\hosts_)

Puis déplacer les scripts présents dans le répertoire _/var/www/html/config_ dans un autre répertoire, par exemple _/opt_ :
```
# mv /var/www/html/config/login_collab1_script.py /opt
# mv /var/www/html/config/login_collab2_script.py /opt
# mv /var/www/html/config/login_manager_script.py /opt
# mv /var/www/html/config/login_admin_script.py /opt
```

Les scripts nécessitent plusieurs paquets/composants afin de fonctionner (Python, Selenium Webdriver ainsi que PhantomJS). Tout d'abord installer Python :
```
# apt-get install python python-pip libfontconfig
```

Puis Selenium :
```
# pip install selenium
```

Enfin PhantomJS :
```
# wget https://bitbucket.org/ariya/phantomjs/downloads/phantomjs-2.1.1-linux-x86_64.tar.bz2
# tar xvjf phantomjs-2.1.1-linux-x86_64.tar.bz2
# cp phantomjs-2.1.1-linux-x86_64/bin/phantomjs /usr/bin/
```

Il est possible d'exécuter les scripts directement et de commencer à attaquer l'application (en accédant à l'application via le navigateur web) :
```
# python -W ignore /opt/login_collab1_script.py &
# python -W ignore /opt/login_collab2_script.py &
# python -W ignore /opt/login_manager_script.py &
# python -W ignore /opt/login_admin_script.py &
```

Il peut être plus judicieux de lancer ses scripts au démarrage de la machine afin de ne pas avoir à les lancer à chaque fois :
```
# vim /etc/systemd/system/rc-local.service
```

Renseigner ceci dans le fichier :
```
[Unit]
Description=/etc/rc.local Compatibility
ConditionPathExists=/etc/rc.local

[Service]
Type=forking
ExecStart=/etc/rc.local start
TimeoutSec=0
StandardOutput=tty
RemainAfterExit=yes
SysVStartPriority=99

[Install]
WantedBy=multi-user.target
```

Créer le fichier _rc.local_ :
```
# vim /etc/rc.local
```

Renseigner ceci :
```
#!/bin/sh -e
#
# rc.local
#
# This script is executed at the end of each multiuser runlevel.
# Make sure that the script will "exit 0" on success or any other
# value on error.
#
# In order to enable or disable this script just change the execution
# bits.
#
# By default this script does nothing.

/usr/bin/python /opt/login_collab1_script.py > /dev/null 2>&1 &
/usr/bin/python /opt/login_collab2_script.py > /dev/null 2>&1 &
/usr/bin/python /opt/login_manager_script.py > /dev/null 2>&1 &
/usr/bin/python /opt/login_admin_script.py > /dev/null 2>&1 &

exit 0
```

Ajouter le script au démarrage du système :
```
# chmod +x /etc/rc.local
# systemctl enable rc-local
```

Redémarrer :
```
# reboot
```

L'installation est maintenant **complète**, l'application est disponible à l'adresse _http://myexpense.fbi.com_


CHANGELOG.md
MyExpense Application
=======================

v1.0.0 (29/04/2018)
======

+ First commit of MyExpense Application

config/config.inc.php
<?php
  // Database Configuration
  $_bdd = array();
  $_bdd['server'] = "127.0.0.1";
  $_bdd['port'] = "3306";
  $_bdd['user'] = "root";
  $_bdd['password'] = "root";
  $_bdd['database'] = "myexpense";

  function createOrRestoreDatabase() {
    global $_bdd;

    $GLOBALS['___mysqli_ston'] = new mysqli($_bdd['server'], $_bdd['user'], $_bdd['password'], "", $_bdd['port']);
    if($GLOBALS['___mysqli_ston']->connect_errno) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->connect_error;
      return;
    }

    if(!($stmt = $GLOBALS['___mysqli_ston']->prepare("DROP DATABASE IF EXISTS " . $_bdd['database']))) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!$stmt->execute()) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!($stmt = $GLOBALS['___mysqli_ston']->prepare("CREATE DATABASE " . $_bdd['database']))) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!$stmt->execute()) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!($stmt = $GLOBALS['___mysqli_ston']->select_db($_bdd['database']))) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!($stmt = $GLOBALS['___mysqli_ston']->prepare("CREATE TABLE site (site_id int(4) NOT NULL AUTO_INCREMENT, city varchar(32), address text, PRIMARY KEY(site_id))"))) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!$stmt->execute()) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!($stmt = $GLOBALS['___mysqli_ston']->prepare("INSERT INTO site VALUES
      ('1', 'Paris', '9 bis rue Dupond Eloise, 78000 VERSAILLES'),
      ('2', 'Rennes', '8 Rue des lilas, 35000 Rennes'),
      ('3', 'Brest', '32 rue de Siam, 29200 Brest')
      "))) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!$stmt->execute()) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!($stmt = $GLOBALS['___mysqli_ston']->prepare("CREATE TABLE user (user_id int(4) NOT NULL AUTO_INCREMENT, username varchar(32) NOT NULL, password varchar(256), role varchar(32), lastname text, firstname text, site_id int(4), mail varchar(128), manager_id int(4), last_connection DATETIME, active tinyint(1), PRIMARY KEY(user_id), UNIQUE(username), UNIQUE(mail), FOREIGN KEY(site_id) REFERENCES site(site_id))"))) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!$stmt->execute()) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!($stmt = $GLOBALS['___mysqli_ston']->prepare("INSERT INTO user VALUES 
      (1, 'afoulon', '" . md5('wq6hblv3') . "', 'Financial approver', 'Foulon', 'Aristide', 1, 'afoulon@futuraBI.fr', 1, NOW(), 1),
      (2, 'pbaudouin', '" . md5('HackMe') . "', 'Financial approver', 'Baudouin', 'Paul', 1, 'pbaudouin@futuraBI.fr', 2, NOW(), 1),
      (3, 'rlefrancois', '" . md5('m4ukhkjq') . "', 'Manager', 'Lefrancois', 'Reynaud', 1, 'rlefrancois@futuraBI.fr', 1, NOW(), 1),
      (4, 'mriviere', '" . md5('6v7j4tj2') . "', 'Manager', 'Riviere', 'Manon', 2, 'mriviere@futuraBI.fr', 2, NOW(), 1),
      (5, 'mnguyen', '" . md5('27sfbdij') . "', 'Manager', 'Nguyen', 'Maximilien', 3, 'mnguyen@futuraBI.fr', 2, NOW(), 1),
      (6, 'pgervais', '" . md5('b98e67ys') . "', 'Collaborator', 'Gervais', 'Placide', 1, 'pgervais@futuraBI.fr', 3, NOW(), 1),
      (7, 'placombe', '" . md5('qayhned2') . "', 'Collaborator', 'Lacombe', 'Philibert', 1, 'placombe@futuraBI.fr', 3, NOW(), 1),
      (8, 'triou', '" . md5('h1rlnifq') . "', 'Collaborator', 'Riou', 'Thierry', 1, 'triou@futuraBI.fr', 3, NOW(), 1),
      (9, 'broy', '" . md5('37f52u21') . "', 'Collaborator', 'Roy', 'Baudouin', 1, 'broy@futuraBI.fr', 3, NOW(), 1),
      (10, 'brenaud', '" . md5('u99hau4d') . "', 'Collaborator', 'Renaud', 'Bernadette', 2, 'brenaud@lrtechnologies.fr', 4, NOW(), 1),
      (11, 'slamotte', '" . md5('fzghn4lw') . "', 'Collaborator', 'Lamotte', 'Samuel', 2, 'slamotte@futuraBI.fr', 4, NOW(), 0),
      (12, 'nthomas', '" . md5('en3dtdjy') . "', 'Collaborator', 'Thomas', 'Ninette', 3, 'nthomas@futuraBI.fr', 5, NOW(), 1),
      (13, 'vhoffmann', '" . md5('qzm8hnmw') . "', 'Collaborateur', 'Hoffmann', 'Victorine', 3, 'vhoffmann@futuraBI.fr', 5, NOW(), 1),
      (14, 'rmasson', '" . md5('tdg33vhe') . "', 'Administrator', 'Masson', 'Rodrigue', 1, 'rmasson@futuraBI.fr', 1, NOW(), 1)
      "))) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!$stmt->execute()) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!($stmt = $GLOBALS['___mysqli_ston']->prepare("CREATE TABLE message (message_id int(4) NOT NULL AUTO_INCREMENT, message text, post_date DATETIME, user_id int(4), PRIMARY KEY(message_id), FOREIGN KEY(user_id) REFERENCES user(user_id))"))) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!$stmt->execute()) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!($stmt = $GLOBALS['___mysqli_ston']->prepare("INSERT INTO message VALUES 
      ('1', 'MyExpense application allow collaborators and managers to report their expenses in order to be reimburesed as quick as possible.', '2018-02-11 10:52:08', '2'),
      ('2', 'Less time wasted than send excel file in a mail!', '2018-02-11 11:23:12', '5'),
      ('3', 'How do I know if my expense report is reimbursed?', '2018-02-11 13:44:43', '12'),
      ('4', 'The status of your expense report will be \" Sent for payment\".', '2018-02-11 14:01:45', '1'),
      ('5', 'Great ! Thank you.', '2018-02-11 16:34:48', '4')
      "))) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!$stmt->execute()) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!($stmt = $GLOBALS['___mysqli_ston']->prepare("CREATE TABLE expense (expense_id int(4) NOT NULL AUTO_INCREMENT, expense_date DATE, amount int(5), comment text, status varchar(32), user_id int(4), order_status int(4), PRIMARY KEY(expense_id), FOREIGN KEY(user_id) REFERENCES user(user_id))"))) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!$stmt->execute()) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!($stmt = $GLOBALS['___mysqli_ston']->prepare("INSERT INTO expense VALUES 
      (1, '2018-02-13', 36, 'Lunch with AI Enterprise Human Ressource', 'Opened', '9', 3),
      (2, '2018-02-14', 584, 'New Ipad, I left mine in the taxi going to the airport.', 'Refused', '5', 5),
      (3, '2018-02-14', 321, 'Train tickets, customer in Nantes.', 'Submitted', '8', 1),
      (4, '2018-02-15', 8, 'Lunch when go to the Welcome Day for new collaborators.', 'Submitted', '7', 1),
      (5, '2018-02-15', 750, 'Plane tickets, Cybersecurity project n°5423545, Toulouse.', 'Opened', '11', 3),
      (6, '2018-02-15', 23, 'Bus and subway tickets (Rennes).', 'Validated', '3', 2),
      (7, '2018-02-16', 333, 'Price of the car rental for the training day (How to improve your soft skills).', 'Opened', '14', 3),
      (8, '2018-02-17', 1200, 'No comment.', 'Sent for payment', '8', 1),
      (9, '2018-02-20', 56, 'Two breakfasts. I was hungry', 'Refused', '8', 5),
      (10, '2018-02-21', 553, 'A new computer.', 'Validated', '4', 2)
      "))) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    if(!$stmt->execute()) {
      $_SESSION['technical_error'] = "Sorry, a technical error has occured : " . $GLOBALS['___mysqli_ston']->error;
      $stmt->close();
      return;
    }

    $_SESSION['message'] = "The database has been created/restored, have fun !";
  }
?>
